scripts/utils/trading_strategy_helpers.js

```diff
@@ -294,15 +302,17 @@ module.exports = function(web3 = web3, artifacts = artifacts) {
     return buildBundledTransaction(transactions)
   }
 
-  const calculateBuyAndSellAmountsFromPrice = function(price, targetToken) {
-    // Sell targetToken for stableToken at price with unlimited orders
-    // Example:
-    // Sell 1 ETH at for 102 DAI (unlimited)
-    // Sell x ETH for max256 DAI
-    // x = max256 / 102
-    // priceFormatted = 102000000000000000000
-    price = price.toFixed(18)
-    const priceFormatted = toErc20Units(price, targetToken.decimals)
+  const calculateBuyAndSellAmountsFromPrice = function(price, stableToken, targetToken, maximalRoundingErrorInPercent = 0.1) {
+    // decimalsForPrice: Decimals of reference price (OWL Price) + stableToken.decimals - targetToken.decimals
+    const decimalsForPrice = 18 + stableToken.decimals - targetToken.decimals
+    if (decimalsForPrice < 0 || decimalsForPrice > 36) {
+      throw "decimalsForPrice is not reasonable. Don't apply the bracket-strategy for this token pair!"
+    }
+    const roundedPrice = price.toFixed(decimalsForPrice)
+    if ((Math.abs(roundedPrice - price) / price) * 100 > maximalRoundingErrorInPercent) {
+      throw "Token decimals will introduce unreasonable rounding error!"
+    }
+    const priceFormatted = toErc20Units(roundedPrice, decimalsForPrice)
```

Comment:
I don't like very much all these restrictions. The number 18 is still around in the code, and it's not very clear why. It doesn't seem to represent OWL, it kind of represent the amount of precision we use for the price, except that we change our precision depending on the difference of the number of decimals of the two tokens. So, if we trade USDC/DAI we have a precision of 6 decimals, while if we trade DAI/USDC we have a precision of 30.
I tried to come up with a suggestion that sets the precision to 18, but I couldn't make BN play together with floats nicely and consistently. I can't say I'm sure this code works as intended, it's likely fine for USDC DAI since the price is about 1, but if there is any token with two decimals I wouldn't be so confident that everything works as intended. Did I misunderstand something? Maybe we could also ask someone else to review the code to hear another opinion.

Reply:
Good concern!

What do you think about my new implementation? 

It feels like we can control the precision pretty easily. I am not too worried about it

Reply:
Much, much nicer! I love the new changes.